internal/command/test.go

```diff
@@ -85,8 +88,72 @@ func (c *TestCommand) Run(rawArgs []string) int {
 		}(),
 	}
 
+	runningCtx, done := context.WithCancel(context.Background())
+	stopCtx, stop := context.WithCancel(runningCtx)
+	cancelCtx, cancel := context.WithCancel(context.Background())
+
+	runner := &TestRunner{
+		command: c,
+
+		Suite:  &suite,
+		Config: config,
+		View:   view,
+
+		CancelledCtx: cancelCtx,
+		StoppedCtx:   stopCtx,
+
+		// Just to be explicit, we'll set the following fields even though they
+		// default to these values.
+		Cancelled: false,
+		Stopped:   false,
+	}
+
 	view.Abstract(&suite)
-	c.ExecuteTestSuite(&suite, config, view)
+
+	go func() {
+		defer logging.PanicHandler()
+		defer done() // We completed successfully.
+		defer stop()
+		defer cancel()
+
+		runner.Start()
+	}()
+
+	// Wait for the operation to complete, or for an interrupt to occur.
+	select {
+	case <-c.ShutdownCh:
+		// Nice request to be cancelled.
+
+		view.Interrupted()
+		runner.Stopped = true
+		stop()
+
+		select {
+		case <-c.ShutdownCh:
+			// The user pressed it again, now we have to get it to stop as
+			// fast as possible.
+
+			view.FatalInterrupt()
+			runner.Cancelled = true
+			cancel()
+
+			// TODO(liamcervante): Should we add a timer here? That would mean
```

Comment:
Maybe there's more info we could get to the user, but a forceful cancellation can (and usually will) always leave untracked resources around because the provider has not been able to return from some number of `ApplyResourceChange` calls. This is a more frequent hazard with user-executed testing, where they too often decide they don't want to wait for whatever reason and panic hit ctrl+c to try and abort.

Reply:
So, I think the implementation right now won't cause this because the forceful cancellation actually only tells it to skip the tidy up, but it still always waits for whatever operation is ongoing to finish (ie. it doesn't pass the forceful cancellation into the terraform context, only the first nice cancel request). I did this because, as you say, we'll almost definitely be left with resources live that Terraform doesn't know about.

If I was to come back and implement the timer here I think we'd start to see the problems, which is why I've just left it as a TODO. 

I'm not sure what the best way to deal with this would be. Potentially we could add a validate/refresh/import option to the test command, and it runs through all the tests and attempts to see if any infrastructure that matches the config exists? That might be very complicated. We already print out all the resources we know exist and haven't been tidied up, I don't know how much more Terraform can actually say.

Reply:
Yeah, without aborting immediately, or at least within a short delay, this isn't going to do much of anything. Most providers don't fully implement cancellation, so this will prevent Terraform from exiting at all without an external signal until the provider times out internally. This has already passed the "there may be data loss" warning, so users should be willing to accept the risk, and now we are probably more likely to see users killing the process entirely to get unstuck.